connect/runtime/src/test/resources/test-plugins/expect-plugin-class-loader/test/plugins/ExpectPluginClassLoader.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package test.plugins;
+
+import java.util.Map;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaAndValue;
+import org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader;
+import org.apache.kafka.connect.runtime.isolation.PluginClassLoader;
+import org.apache.kafka.connect.storage.Converter;
+import org.junit.Assert;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Verify that a PluginClassLoader is being used to load this class at all times.
+ */
+public class ExpectPluginClassLoader implements Converter {
+
+    private static final Logger log = LoggerFactory.getLogger(ExpectPluginClassLoader.class);
+
+    private static final ClassLoader staticClassloader;
+
+    static {
+        log.info("Starting static initialization");
+        staticClassloader = Thread.currentThread().getContextClassLoader();
+        log.info("Finished static initialization");
+    }
+
+    {
```

Comment:
I'm wondering if it might be possible (and better style) to limit the activity of this class to only collecting information as opposed to collecting and verifying it. Performing assertions during class initialization has the potential to obscure the stack traces they generate which, especially if messages are added to those assertions, are generally quite useful. Plus, it's a little out of the ordinary to perform assertions in non-test-class code; personally, I think it'd be more readable to have them all in one place (e.g., `PluginsTest` and/or `DelegatingClassLoaderTest` as opposed to a mix of those test classes and this utility class).
I'm thinking this class can retain its `staticClassLoader` field (although made `public` instead of `private`) and `Thread.currentThread.getContextClassLoader()` and `ExpectPluginClassLoader.class.getClassLoader()` can be stored as `public final` instance variables, then assertions against them can be performed by whatever test class decides to load them.

Also, it'd probably be possible to eliminate the `AlwaysSucceed` plugin and just use this one in its place if it can be loaded without any automatic assertions being performed.

If you decide to go with one or both of these suggestions (absorbing the `AlwaysSucceed` plugin into this one, and removing the automatic assertions in favor of exposing collected data to test classes) it'd probably be good to rename this class too; I'd suggest something like `RecordClassLoaders` but I'm sure there's a better name out there so feel free to change it if desired.

Reply:
I've had problems debugging when exceptions are thrown during static initialization, and fixed this by pushing all of the asserts into instance initialization. The stacktraces from exceptions thrown from instance initialization seem fine for debugging. I agree that the test plugins code is pretty far from `TestPlugins` or `PluginsTest`, but maybe this can be addressed by some better comments.

It does seem much simpler to just have one test plugin that samples its classloader and then delivers it to the actual test code once it's loaded, but i'm not sure how to implement this. How do we retrieve the result after the object has been created?
1. Reflectively call a getter method, but this might be hard to follow during debugging.
2. Add a superclass with the getter, but this is a non-option since I would have to add this to the list of classes that are recognized as plugins in the production code.
3. Tunnel the result through an `Object` reference, such as `SchemaAndValue::value`, but this is really non-idiomatic.

Reflection seems like the easiest here, but should I really be adding more reflection here?

Reply:
@C0urante and I discussed having these test plugins implement a test interface, and then perform a cast after instantiating the plugin. This seems like the cleanest solution, so I'll go back and implement this change.